exec/command.go

```diff
@@ -0,0 +1,186 @@
+type ExitCodeError struct {
+	*Result
+}
+
+// Error returns the error description.
+func (e ExitCodeError) Error() string {
+	return fmt.Sprintf("exec: running '%s' in directory '%s' exited with code %d, expected 0, output: '%s'",
+		e.Command, e.Dir, e.ExitCode, e.Output)
+}
+
+// Cmd represents a command that can be run.
+type Cmd struct {
+	path string
+	args []string
+
+	dir           string
+	debugfFn      func(format string, v ...interface{})
+	debugfPrefix  string
+	expectSuccess bool
+}
+
+// Command returns a new Cmd struct.
+// If name contains no path separators, Command uses LookPath to
+// resolve name to a complete path if possible. Otherwise it uses name directly
+// as Path.
+// By default a command is run in the current working directory.
+func Command(name string, arg ...string) *Cmd {
+	return &Cmd{
+		path:         name,
+		args:         arg,
+		dir:          ".",
+		debugfFn:     DefaultDebugfFn,
+		debugfPrefix: DefaultDebugPrefix,
+	}
+}
+
+// ShellCommand executes a command in sh shell.
+func ShellCommand(cmd string) *Cmd {
+	return Command("sh", "-c", cmd)
+}
+
+// Directory changes the directory in which the command is executed.
+func (c *Cmd) Directory(dir string) *Cmd {
+	c.dir = dir
+	return c
+}
+
+// DebugfFunc sets the debug function for the command. It accepts a
+// printf-style printf function and call it for every line that the command
+// prints to STDOUT and STDERR when it's run.
+func (c *Cmd) DebugfFunc(fn func(format string, arg ...interface{})) *Cmd {
+	c.debugfFn = fn
+	return c
+}
+
+// DebugfPrefix sets a prefix that is prepended to the message that is passed to the Debugf function.
+func (c *Cmd) DebugfPrefix(prefix string) *Cmd {
+	c.debugfPrefix = prefix
+	return c
+}
+
+// ExpectSuccess if called, Run() will return an error if the command did not
+// exit with code 0.
+func (c *Cmd) ExpectSuccess() *Cmd {
+	c.expectSuccess = true
+	return c
+}
+
+func cmdString(cmd *exec.Cmd) string {
+	// cmd.Args[0] contains the command name, cmd.Path the absolute command path,
+	// omit cmd.Args[0] from the string
+	if len(cmd.Args) > 1 {
+		return fmt.Sprintf("%s %v", cmd.Path, strings.Join(cmd.Args[1:], " "))
+	}
+
+	return cmd.Path
+}
+
+// Result describes the result of a run Cmd.
+type Result struct {
+	Command  string
+	Dir      string
+	Output   []byte
+	ExitCode int
+}
+
+// StrOutput returns Result.Output as string.
+func (r *Result) StrOutput() string {
+	return string(r.Output)
+}
+
+// ExpectSuccess if the ExitCode in Result is not 0, the function returns an
+// ExitCodeError for the execution.
+func (r *Result) ExpectSuccess() error {
+	if r.ExitCode == 0 {
+		return nil
+	}
+
+	return ExitCodeError{Result: r}
+}
+
+// Run executes the command.
+func (c *Cmd) Run() (*Result, error) {
+	cmd := exec.Command(c.path, c.args...)
+	cmd.Dir = c.dir
+
+	outReader, err := cmd.StdoutPipe()
+	if err != nil {
+		return nil, err
+	}
+	cmd.Stderr = cmd.Stdout
```

Comment:
For this PR I think this is ok. In general I think joining stdout and -err is too restrictive.

Reply:
Yes, maybe we will need in the future to distinguish between stdout and stderr.